### ISSUE TITLE
React-Standalone with Scale & Events

### ISSUE THREAD

**Issue by SebastianGode** [telekom/scale#642]
I'm trying to modify an existing page and as a result need to use react standalone and include Scale&React through plain html:

```
<link rel="stylesheet" href="node_modules/@telekom/scale-components/dist/scale-components/scale-components.css">
<script type="module" src="node_modules/@telekom/scale-components/dist/scale-components/scale-components.esm.js"></script>
<script src="https://unpkg.com/babel-standalone@6/babel.min.js"></script>
<script src="https://unpkg.com/react@17/umd/react.development.js" crossorigin></script>
<script src="https://unpkg.com/react-dom@17/umd/react-dom.development.js" crossorigin></script>
```
The issue with that approach is that Events won't work together with React. You have provided an wrapper for that @telekom/scale-components-react but the issue is that I have no clue how to include that in plain html as I don't know which file to include. Same goes for reactify-ws. Is there a way to do that?

**Comment by acstll** [comment]
Hello @SebastianGode 

I'm deeply sorry it took us so long to get back to you.

You were right, there is no way to use the wrappers package without a build step, because it exports "esm-compatible" CommonJS only. Which is a shame.

The [`reactify-wc`](https://github.com/BBKolton/reactify-wc) package, that would also serve to work around React's limitations with custom events, also exports only CJS and ESM, but the ES module tries to import React, so it needs a build step anyways.

The solution, as you mentioned already in #639, is to offer an UMD version for the wrapper package. (A PR for a CDN-friendly `reactify-wc` would also be cool). We're adding this to our backlog.

You probably know this, but there's still a way to handle custom events in React, which is a bit cumbersome, but valid anyways: using plain old `addEventListener` with references to the DOM elements. I'm sharing a snippet of a working example:

```html 
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <link rel="stylesheet" href="https://unpkg.com/@telekom/scale-components/dist/scale-components/scale-components.css">
  <script type="module" src="https://unpkg.com/@telekom/scale-components/dist/scale-components/scale-components.esm.js"></script>
  <script src="https://unpkg.com/babel-standalone@6/babel.min.js"></script>
  <script src="https://unpkg.com/react@17/umd/react.development.js" crossorigin></script>
  <script src="https://unpkg.com/react-dom@17/umd/react-dom.development.js" crossorigin></script>
  <title>Document</title>
</head>
<body>
  <div id="app"></div>

  <script type="text/babel">
    function App() {
      const modal = React.useRef(null)
      const trigger = React.useRef(null)
      const [opened, set] = React.useState(false)
      const handleClick = () => {
        set(true)
      }
      const handleClose = () => {
        set(false)
        trigger.current.setFocus()
      }
      React.useEffect(() => {
        modal.current.addEventListener('scale-close', handleClose)
        return () => {
          modal.current.removeEventListener('scale-close', handleClose)
        }
      }, [])

      return (
        <div>
          <h1>It (kinda) works!</h1>
          <scale-button ref={trigger} onClick={handleClick}>Hi</scale-button>
          <scale-modal heading="Hello!" ref={modal} opened={opened}>
            <p>👋</p>
          </scale-modal>
        </div>
      )
    }

    ReactDOM.render(<App />, document.getElementById('app'))
  </script>
</body>
</html>
```

Sorry once more, and thanks for bringing this one up.

**Comment by christopherGdynia** [comment]
Hey, can you these unpkg packages to the Storybook? :D

**Comment by acstll** [comment]
Hey @SebastianGode I hope things are good!

We still need to add the UMD build (it will come), but… I wanted to share something very interesting:

Support for custom elements landed in React in the `@experimental` dist. (I guess you install that like `npm i react@experimental react-dom@experimental`). Event bindings there should work without any wrappers 🤞 

Please check it out here: https://github.com/facebook/react/issues/11347#issuecomment-988970952

Maybe this is something you want to test. Any feedback will be appreciated 🙏

**Comment by SebastianGode** [comment]
Hey @acstll 
We still don't have a UMD build, right?
Because currently Scale is still completely unusable on non-Node projects.

**Comment by acstll** [comment]
Hello @SebastianGode

we sadly couldn't manage to prioritise this one: adding an UMD build to the React wrapper package (namely `@telekom/scale-components-react`).

We will try and make this possible, it should be (we cannot say when, but I will give it a shot myself this week).

I would like to point out that, and I hope you agree, the statement "currently Scale is still completely unusable on non-Node projects" is not true. It's completely possible as demonstrated above with the following snippet:

```html
  <link rel="stylesheet" href="https://unpkg.com/@telekom/scale-components/dist/scale-components/scale-components.css">
  <script type="module" src="https://unpkg.com/@telekom/scale-components/dist/scale-components/scale-components.esm.js"></script>
```

[`<script type=module>` is already widely supported everywhere.](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script#browser_compatibility)

The problem is exclusive to the React wrapper package, for which there's no way to use without a build/bundle step. And because of the very nature of React, I would consider using it without a build step an edge case. (However, I'm personally very much in favor of avoiding build steps whenever possible.)

In summary, we're gonna try and add a UMD output for `@telekom/scale-components-react` only, hoping it will work together with the default ESM build of the main library.

Again I'm very sorry for being unable to provide a proper solution for your problem during all this time. I hope you understand issues going stale sometimes is a common problem all open sourced project suffer from, e.g. some of us have been waiting for custom element support in React for very long (https://github.com/facebook/react/issues/11347#issue-267902244).

**Comment by SebastianGode** [comment]
Hey @acstll 
Thanks for the answer!
Using an external CDN isn't possible on some pages due to the nature of it being external.
I tried compiling scale by myself (and after some trouble also got it to compile), but the amount of files produced is in my opinion also not a really good solution. Yes, you can include it too with only two lines (like with unpkg), but in case you put it in a git repo you will add tons of files. Compared to e.g. react which only has a single file you need (https://unpkg.com/browse/react@18.2.0/umd/) this is definitely not a great solution currently.

Even without using React, but Scale alone on a static pure-JS webpage, it will require a lot of files and is not really great to include.

**Comment by acstll** [comment]
My pleasure. Thanks for the valuable feedback.

You're completely right about the single file/many files situation. There's tons of imports here: `https://unpkg.com/@telekom/scale-components@3.0.0-beta.56/dist/scale-components/scale-components.esm.js`. And you're also right in that a CDN being required is not convenient and sometimes a blocker.

It's fair to say that currently Scale is unfriendly (or unwelcoming) for non-Node projects 👍

**Comment by acstll** [comment]
My pleasure.

Thanks for the valuable feedback. You're completely right about the single file/many files situation. There's tons of imports here: `https://unpkg.com/@telekom/scale-components@3.0.0-beta.56/dist/scale-components/scale-components.esm.js`. And you're also right in that a CDN being required is not convenient.

**Comment by SebastianGode** [comment]
@acstll 
Any news regarding a UMD build with a single JS and CSS file?

**Comment by ayeshaa-m** [comment]
The umd build creation is done and it will be available in upcoming release. https://github.com/telekom/scale/issues/639#issuecomment-1603946987

**Comment by acstll** [comment]
@SebastianGode, as Ayesha mentioned, there will be an UMD build in the upcoming release for the `@telekom/scale-components-react` package. (Thank you @ayeshaa-m)

I wanted to let you know that I also did some tests around the idea of making Scale friendlier for non-Node projects.

One the one hand, an UMD build of the main package seems far from optimal. The total size of the single JS file would be around 2.5 MB or 1.5 MB minified (don't know about gzip, but that's big enough for me). Including all components and the +200 icons. Also I noticed some runtime errors for some components which I don't get with the default lazy-loading mechanism. Plus, with this UMD build, one would have to call a function for every component needed, e.g. `window.ScaleComponents.defineCustomElementScaleButton()`.

On the other hand, I think providing a ZIP file for every release with the relevant files shouldn't be hard. We could automate this in the release process and make it available in the GitHub releases page. Only the [`dist/scale-components/`](https://unpkg.com/browse/@telekom/scale-components@3.0.0-beta.135/dist/) folder would be needed. This folder has:
- the styles in [`scale-components.css`](https://unpkg.com/browse/@telekom/scale-components@3.0.0-beta.135/dist/scale-components/scale-components.css)
- the [`scale-components.esm.js`](https://unpkg.com/browse/@telekom/scale-components@3.0.0-beta.135/dist/scale-components/scale-components.esm.js)  ES module that lazy-loads all components
- all component modules as separated JS files
- fonts

(Currently there's an empty `index.esm.js` file and a JSON file inside `components/icons` which is not needed, but we can clean that up.)

With this, in a none-Node project, one could update Scale by just replacing this one folder. Loading would mean:

```html
<link rel="stylesheet" href="assets/scale-components/scale-components.css">
<script type="module" src="assets/scale-components/scale-components.esm.js"></script>
```

Please let us know what you think.

Thanks again for your interest and feedback.

**Comment by SebastianGode** [comment]
Hey @acstll 
Thanks for the reply and the work!

> One the one hand, an UMD build of the main package seems far from optimal. The total size of the single JS file would be around 2.5 MB or 1.5 MB minified (don't know about gzip, but that's big enough for me). Including all components and the +200 icons. Also I noticed some runtime errors for some components which I don't get with the default lazy-loading mechanism. Plus, with this UMD build, one would have to call a function for every component needed, e.g. window.ScaleComponents.defineCustomElementScaleButton().

I have thought about that quite some time now. I'm not sure what's ideal here, because you are right with the big bundle size. A minified version with gzip or brotli compression should hopefully be less than 1MB but that's still really large. What takes up so much space? All the icons? Maybe a "light" version or something like that would be possible. Having the icons separately would also be okay in my opinion as then they would only be loaded if you use them. Also when Lazy-Loading the multiple JS files, shouldn't the overall size on page load not be the same? Or will some JS files only be loaded if you use the corresponding component?

> On the other hand, I think providing a ZIP file for every release with the relevant files shouldn't be hard. We could automate this in the release process and make it available in the GitHub releases page. Only the [dist/scale-components/](https://unpkg.com/browse/@telekom/scale-components@3.0.0-beta.135/dist/) folder would be needed.

I think it's a good idea to bundle that as a ZIP file for each release.
However I'd also love to have a UMD build even with the large bundle size as even the dist folder has still a lot of files. Commiting them all to a git repository (and replacing them in case of an update) would still not be ideal. To have the option to at least get an single, minified CSS and JS file would be great, even if it's large and not really used that often as a result.

**Comment by SebastianGode** [comment]
Hey @acstll 
When can we expect a Zip with the dist flder on each release? The current Beta tags don't have it yet.

**Comment by tshimber** [comment]
hi @SebastianGode ! please tell if the issue is still relevant. the one with zip seems to be solved
